File: saki-api/src/saki_api/models/permission.py
"""
Permission system models for RBAC (Role-Based Access Control).

Includes:
- Role enums (GlobalRole, ResourceRole)
- Permission enum
- RolePermission mapping table
- DatasetMember table for resource-level permissions
"""

from datetime import datetime
from enum import Enum
from typing import Optional, TYPE_CHECKING

from saki_api.models.base import TimestampMixin, UUIDMixin
from sqlmodel import Field, SQLModel, Relationship

if TYPE_CHECKING:
    from saki_api.models.user import User
    from saki_api.models.dataset import Dataset


# ============================================================================
# Role Enums
# ============================================================================

class GlobalRole(str, Enum):
    """全局角色：系统级角色"""
    SUPER_ADMIN = "super_admin"
    ADMIN = "admin"
    ANNOTATOR = "annotator"
    VIEWER = "viewer"


class ResourceRole(str, Enum):
    """资源角色：数据集/项目级别的角色"""
    OWNER = "owner"
    MANAGER = "manager"
    ANNOTATOR = "annotator"
    REVIEWER = "reviewer"
    VIEWER = "viewer"


# ============================================================================
# Permission Enum
# ============================================================================

class Permission(str, Enum):
    """权限枚举：定义所有可用的权限"""
    # 用户管理
    USER_CREATE = "user:create"
    USER_READ = "user:read"
    USER_UPDATE = "user:update"
    USER_DELETE = "user:delete"
    USER_MANAGE_ROLES = "user:manage_roles"

    # 数据集全局权限
    DATASET_CREATE = "dataset:create"
    DATASET_READ_ALL = "dataset:read_all"
    DATASET_UPDATE_ALL = "dataset:update_all"
    DATASET_DELETE_ALL = "dataset:delete_all"

    # 数据集资源权限
    DATASET_READ = "dataset:read"
    DATASET_UPDATE = "dataset:update"
    DATASET_DELETE = "dataset:delete"
    DATASET_MANAGE_MEMBERS = "dataset:manage_members"
    DATASET_UPLOAD = "dataset:upload"
    DATASET_EXPORT = "dataset:export"

    # 样本权限（继承自数据集）
    SAMPLE_READ = "sample:read"
    SAMPLE_UPDATE = "sample:update"
    SAMPLE_DELETE = "sample:delete"

    # 标注权限（继承自数据集）
    ANNOTATION_READ = "annotation:read"
    ANNOTATION_MODIFY = "annotation:modify"  # 合并create/update/delete
    ANNOTATION_REVIEW = "annotation:review"

    # 系统配置
    SYSTEM_CONFIG = "system:config"


# ============================================================================
# Role-Permission Mapping
# ============================================================================

class RolePermission(SQLModel, table=True):
    """
    角色-权限关联表：定义每个角色拥有哪些权限
    
    资源类型：
    - None: 全局权限
    - "dataset": 数据集资源权限
    """
    __tablename__ = "role_permission"

    role: str = Field(primary_key=True, description="角色名称（GlobalRole或ResourceRole）")
    permission: Permission = Field(primary_key=True, description="权限")
    resource_type: Optional[str] = Field(
        default=None,
        description="资源类型：None表示全局权限，'dataset'表示数据集资源权限"
    )


# ============================================================================
# Dataset Member (Resource-level permissions)
# ============================================================================

class DatasetMember(SQLModel, table=True):
    """
    数据集成员表：用户-数据集-角色关联
    
    用于实现数据集级别的细粒度权限控制
    """
    __tablename__ = "dataset_member"

    dataset_id: str = Field(
        foreign_key="dataset.id",
        primary_key=True,
        description="数据集ID"
    )
    user_id: str = Field(
        foreign_key="user.id",
        primary_key=True,
        description="用户ID"
    )
    role: ResourceRole = Field(
        description="用户在该数据集中的角色"
    )
    created_at: datetime = Field(
        default_factory=datetime.utcnow,
        description="添加时间"
    )
    created_by: str = Field(
        foreign_key="user.id",
        description="谁添加的这个成员"
    )

    # Relationships
    dataset: "Dataset" = Relationship(back_populates="members")
    user: "User" = Relationship(
        back_populates="dataset_memberships",
        sa_relationship_kwargs={
            "foreign_keys": "[DatasetMember.user_id]"
        }
    )
    creator: "User" = Relationship(
        sa_relationship_kwargs={
            "foreign_keys": "[DatasetMember.created_by]"
        }
    )


# ============================================================================
# Schema Models
# ============================================================================

class DatasetMemberCreate(SQLModel):
    """创建数据集成员的请求模型"""
    user_id: str = Field(description="用户ID")
    role: ResourceRole = Field(description="角色")


class DatasetMemberRead(SQLModel):
    """读取数据集成员的响应模型"""
    dataset_id: str
    user_id: str
    role: ResourceRole
    created_at: datetime
    created_by: str
    # 可选：包含用户信息
    user_email: Optional[str] = None
    user_full_name: Optional[str] = None


class DatasetMemberUpdate(SQLModel):
    """更新数据集成员的请求模型"""
    role: ResourceRole = Field(description="新角色")

File: saki-web/src/types/permission.ts
/**
 * Permission System Types
 *
 * Defines types for the RBAC permission system.
 */

// ============================================================================
// Enums
// ============================================================================

export type RoleType = 'system' | 'resource';
export type ResourceType = 'dataset' | 'project';
export type Scope = 'all' | 'assigned' | 'self';

/**
 * Permission object representation
 *
 * Scope hierarchy (from highest to lowest):
 * - all: System-level permissions, only for system roles
 * - assigned: Resource-level permissions, for resource role members
 * - self: Resource-level permissions, for own resources/annotations
 */
export interface Permission {
    target: string;  // Resource type (e.g., 'user', 'dataset', '*')
    action: string;  // Action (e.g., 'create', 'read', '*')
    scope: Scope;    // Permission scope: 'all' | 'assigned' | 'self'
}

/**
 * Parse permission string to Permission object
 * Format: "target:action:scope" or "target:action" (defaults to 'assigned')
 *
 * @example
 * parsePermission("user:create:all") => { target: "user", action: "create", scope: "all" }
 * parsePermission("dataset:*:assigned") => { target: "dataset", action: "*", scope: "assigned" }
 * parsePermission("annotation:read") => { target: "annotation", action: "read", scope: "assigned" }
 */
export function parsePermission(permissionStr: string): Permission {
    const parts = permissionStr.split(':');
    if (parts.length < 2) {
        throw new Error(`Invalid permission format: ${permissionStr}`);
    }

    const target = parts[0];
    const action = parts[1];
    const scope = (parts[2] as Scope) || 'assigned';

    // Validate scope
    if (scope !== 'all' && scope !== 'assigned' && scope !== 'self') {
        throw new Error(`Invalid scope: ${scope}. Must be 'all', 'assigned', or 'self'`);
    }

    return {target, action, scope};
}

/**
 * Convert Permission object back to string
 */
export function permissionToString(permission: Permission): string {
    return `${permission.target}:${permission.action}:${permission.scope}`;
}

// ============================================================================
// Role Types
// ============================================================================

export interface RoleInfo {
    id: string;
    name: string;
    displayName: string;
    description: string;
    color: string;
    isSupremo: boolean;
}

export interface Role {
    id: string;
    name: string;
    displayName: string;
    description?: string;
    type: RoleType;
    parentId?: string;
    isSuperAdmin: boolean;
    isSystem: boolean;
    isDefault: boolean;
    isSupremo: boolean;
    sortOrder: number;
    color: string;
    createdAt: string;
    updatedAt?: string;
    permissions: RolePermission[];
}

export interface RolePermission {
    id: string;
    permission: string;
    conditions?: Record<string, any>;
}

export interface RoleCreate {
    name: string;
    displayName: string;
    description?: string;
    type: RoleType;
    color: string;
    parentId?: string;
    permissions: RolePermissionCreate[];
}

export interface RolePermissionCreate {
    permission: string;
    conditions?: Record<string, any>;
}

export interface RoleUpdate {
    displayName?: string;
    description?: string;
    color?: string;
    parentId?: string;
    sortOrder?: number;
    permissions?: RolePermissionCreate[];
}

// ============================================================================
// User Role Types
// ============================================================================

export interface UserSystemRole {
    id: string;
    userId: string;
    roleId: string;
    assignedAt: string;
    assignedBy?: string;
    expiresAt?: string;
    roleName?: string;
    roleDisplayName?: string;
}

export interface UserSystemRoleAssign {
    roleId: string;
    expiresAt?: string;
}

// ============================================================================
// Resource Member Types
// ============================================================================

export interface ResourceMember {
    id: string;
    resourceType: ResourceType;
    resourceId: string;
    userId: string;
    roleId: string;
    createdAt: string;
    createdBy?: string;
    updatedAt?: string;
    userEmail?: string;
    userFullName?: string;
    userAvatarUrl?: string;
    roleName?: string;
    roleDisplayName?: string;
    roleColor?: string;
}

export interface ResourceMemberCreate {
    userId: string;
    roleId: string;
}

export interface ResourceMemberUpdate {
    roleId: string;
}

// ============================================================================
// Permission Response Types
// ============================================================================

/**
 * System-level permissions (from system roles)
 */
export interface SystemPermissions {
    userId: string;
    systemRoles: RoleInfo[];
    permissions: string[];  // Permission strings, will be parsed to Permission objects
    isSuperAdmin: boolean;
}

/**
 * Resource-specific permissions (from resource roles)
 */
export interface ResourcePermissions {
    resourceRole?: RoleInfo;
    permissions: string[];  // Permission strings, will be parsed to Permission objects
    isOwner: boolean;
}

// ============================================================================
// Permission Store Types
// ============================================================================

export interface ResourcePermissionCache {
    resourceType: string;
    resourceId: string;
    role?: RoleInfo;
    permissions: string[];  // Permission strings, will be parsed to Permission objects
    isOwner: boolean;
    fetchedAt: number;
}

/**
 * Permission index structure for fast lookup
 *
 * Structure:
 * - Map<target, Map<action, Set<scope>>>
 *
 * This allows O(1) lookup instead of O(n) iteration
 */
export type PermissionIndex = Map<string, Map<string, Set<Scope>>>;

// ============================================================================
// Permission Constants
// ============================================================================

export const Permissions = {
    // System
    SYSTEM_MANAGE: 'system:manage:all',

    // User management
    USER_CREATE: 'user:create:all',
    USER_READ: 'user:read:all',
    USER_UPDATE: 'user:update:all',
    USER_DELETE: 'user:delete:all',
    USER_MANAGE: 'user:manage:all',

    // Role management
    ROLE_CREATE: 'role:create:all',
    ROLE_READ: 'role:read:all',
    ROLE_UPDATE: 'role:update:all',
    ROLE_DELETE: 'role:delete:all',

    // Dataset
    DATASET_CREATE: 'dataset:create:all',
    DATASET_READ: 'dataset:read:assigned',
    DATASET_UPDATE: 'dataset:update:assigned',
    DATASET_DELETE: 'dataset:delete:assigned',
    DATASET_ASSIGN: 'dataset:assign:assigned',
    DATASET_EXPORT: 'dataset:export:assigned',
    DATASET_IMPORT: 'dataset:import:assigned',

    // Sample
    SAMPLE_READ: 'sample:read:assigned',
    SAMPLE_CREATE: 'sample:create:assigned',
    SAMPLE_UPDATE: 'sample:update:assigned',
    SAMPLE_DELETE: 'sample:delete:assigned',

    // Annotation
    ANNOTATION_READ: 'annotation:read:assigned',
    ANNOTATION_CREATE: 'annotation:create:assigned',
    ANNOTATION_UPDATE: 'annotation:update:assigned',
    ANNOTATION_DELETE: 'annotation:delete:assigned',
    ANNOTATION_REVIEW: 'annotation:review:assigned',

    // Annotation - self scope
    ANNOTATION_READ_SELF: 'annotation:read:self',
    ANNOTATION_UPDATE_SELF: 'annotation:update:self',
    ANNOTATION_DELETE_SELF: 'annotation:delete:self',
} as const;

export type PermissionKey = keyof typeof Permissions;
